dash-high-level-design (#33)
Azure/DASH: Documentation/dash-hld.md

The overall requirement is to **optimize network SMART Programmable Technologies performance**, and **leverage commodity hardware technology** to achieve **10x or even 100x stateful connection performance**.
- With the help of network hardware vendors, create an open forum that capitalizes on the use of **programmable networking hardware** including SmartNICs, SmartToRs, SmartAppliances. 
- Optimize **stateful L4** performance and connection scale by 10x or even 100x when compared to implementations that make extensive use of software. As host networking in the cloud is performed at L4 the resulting performance improvements should be truly significant.
- Microsoft Azure will integrate and deploy DASH solutions to ensure that scale, monitoring, reliability, availability and constant innovation are proven and hardened. It should be noted that innovations for **in-service software upgrades** (ISSU) and **high availability** (HA) are key tenets of the DASH charter.   

## Architecture

SONiC is structured into various containers that communicate through multiple logical databases via a shared Redis instance. DASH will make use of the SONiC infrastructure as shown in the figure below.

> 

Reviewer: **SDN Controller**
SAI is the Southbound interface facing the hardware. 
SONiC will be calling SAI.  
SAI is not exposed as gRPC, and is not in gNMI format.  
We need to have a method to call the card remotely; we are going to use gRPC.  
The idea is to create abstraction layer on top of SAI interface (that suppliers have), with an agent translating and driving the goal sate and calling the SAI API to configure the hardware.
The NB API is a way to control the card through the control-plane.  

The SDN controller is the overlay that we build on top of the physical layer of the infrastructure.  From the point of view of the SDN control plane, when a customer creates an operation (ex: VNET creation) in our user interface, the calls come in to the control plane to allocate the resources, placement management, decide on capacity, etc...  Our control plane needs to call the hardware and the software for configuration.  The API calls we will be calling are the 'NorthBound' interface API's for customers to control the behavior.  gNMI is the open format we are extending with additional set of APIs on top of the gNMI to perform actions related to SDN.  We will have the API to work with the abstraction layer for SDN.  We will have a NB interface for creating and deleting the Elastic Network Interface (ENI), create Access Lists (ACLs) and specify the rules, establish the routing and mappings, encapsulate the traffic; all of the APIs will allow us (from the NB) to control this behavior.  There will be an Agent we will develop and publish to the Community (with help) to provide the translation (NB APIs exposed over gRPC to be called from the control plane).  Each DPU via gNMI will expose its own management ID/IP (depending upon availability of IPv4 address space - we may have to optimize).  _In the case of a Smart Switch_, the SDN controller management service will identify the specific card that needs to be programmed, it will 'call' the card via the exposed interface.  No complex logic will run on the switches (switches do not have a top-level view of other/neighboring switches in the infrastructure).  

For High Availability, the SDN controller selects the pair of cards that will have the same goal-state.  It will always be scoping the control-plane state.  For example:  the SDN controller will call gNMI on Card A and Card B to configure ENIs, update mappings for communication, etc..  The only requirement on the card from the HA perspective is for the cards to setup a channel between themselves for flow synchronization.  We will not be replicating flows from the control-plane.  We will leave this open for the supplier to provide the most efficient method to synchronize this.

Author: @KrisNey-MSFT and @chrispsommers I incorporated your writing into the Architecture section and created a layered high-level diagram that Kristina sanity-checked. Obviously, all of it can be improved. But I think we are getting there.

Author: Making @KrisNey-MSFT and @chrispsommers change in comment to keep a record. I did add the relevant info into the Architectures section in different format. 
## SDN Architecture
The SDN controller is primarily responsible for controlling the DASH overlay services. The traditional SONiC stack is used to manage the underlay (L3 routing) and hardware platform. The SDN controller communicates to a DASH device through a gNMI endpoint served by a new DASH SDN agent running inside a new SONiC DASH container.  Microsoft will spearhead the development of the SDN Agent and solicit Community participation. This container translates SDN configuration modeled in gNMI into SONiC APPL_DB config objects. The SONiC `orchagent` will be enhanced to translate these objects into SAI_DB objects, including the new DASH-specific SAI objects. An enhanced `syncd` will then configure the dataplane using the vendor-specific SAI library. 
The SDN controller controls the overlay built on top of the physical layer of the infrastructure.  From the point of view of the SDN control plane, when a customer creates an operation (ex: VNET creation) in a portal, the controller allocates the resources, placement management, capacity, etc. via the 'NorthBound' interface APIs.
gNMI schema (TODO - provide URL) will be created for managing DASH services: the Elastic Network Interface (ENI), Access Control Lists (ACLs), routing and mappings, encapsulations; etc.  MSFT will spearhead the development of an SDN Agent and solicit Community participation.
### Multiple DPUs in a Device
Each DPU provides a gNMI endpoint for SDN control through a unique IP address. An appliance or smart switch containing multiple DPUs therefore contains multiple gNMI endpoints for SDN control, and the controller treats each DPU as a separate entity. To conserve IPv4 addresses, such an appliance or switch _might_ contain a proxy (NAT) function to map a single IP address:port combination into multiple DPU endpoints with unique IPv4 addresses.  No complex logic will run on the switches (switches do not have a top-level view of other/neighboring switches in the infrastructure). 
## SDN and DPU High-Availability
For High Availability, the SDN controller selects the pair of cards and configures them identically.  The only requirement on the card from the HA perspective is for the cards to setup a channel between themselves for flow synchronization.  The synchronization mechanism is left for vendors to define and implement.  See <link to HA document>.

Author: Content included.

> [!NOTE] 
> **The figures shown are a work in progress**

> 

Reviewer: Place Representative Region & DC Architecture here (Kristina)

### DASH NOS single DPU on NIC

![dash-single-dpu-architecture](images/hld/dash-single-dpu-architecture.svg)

> 

Reviewer: Could we turn the DASH user space container to match the yellow in the sai-api-dash yellow color (to show that it has changed)?  In all 3 of the drawings?

Author: Done

### DASH appliance architecture

#### High level architecture

![dash-high-level-appliance](images/hld/dash-high-level-appliance.svg)

#### Low level architecture

![dash-appliance-architecture](images/hld/dash-appliance-architecture.svg)

### DASH smart switch architecture

#### High level architecture

![dash-high-level-smart-switch](images/hld/dash-high-level-smart-switch.svg)

#### Low level architecture

![dash-smart-switch-architecture](images/hld/dash-smart-switch-architecture.svg)

## High-Level Design

> 

Reviewer: Change required

Author: Done

DASH uses an **enhanced syncd** to configure the dataplane using the vendor-specific SAI library.

### ASIC Drivers
The term "ASIC Drivers" is borrowed from traditional SONiC and SAI, where a datacenter switch was implemented almost entirely inside an ASIC (or multiple ASICs). These devices are programmed using a vendor Software Development Kit (SDK) which includes device drivers, kernel modules, etc.

A contemporary DASH "SmartNIC" may consist of many complex hardware components including multi-core System On A Chip (SoC) ASICs, and the associated software. For simplicity, the software for such systems which interfaces to the SAI layer is collectively called the "ASIC driver." More importantly, the vendor SAI library will hide all details and present a uniform interface.

### DASH capable ASICs
These comprise the main dataplane engines and are the core of what are variously called SmartNICs, DPUs, IPUs, NPUS, etc. The actual cores may be ASICs, SoCs, FPGAs, or some other high-density, performant hardware.

## Detailed architectures

### DASH NOS single DPU on NIC

![dash-single-dpu-architecture](images/hld/dash-single-dpu-architecture.svg)

> 

Author: Fixed spelling error. We need to decide about the simplification.

### DASH appliance architecture

> #### High level architecture

Reviewer: ```suggestion
#### DASH container state interactions representation
```

Author: Done

![dash-high-level-appliance](images/hld/dash-high-level-appliance.svg)

#### Low level architecture

![dash-appliance-architecture](images/hld/dash-appliance-architecture.svg)

### DASH smart switch architecture

#### High level architecture

![dash-high-level-smart-switch](images/hld/dash-high-level-smart-switch.svg)

#### Low level architecture

> ![dash-smart-switch-architecture](images/hld/dash-smart-switch-architecture.svg)

Author: Done.

In the **sync-d container** we will add **sai api DASH** (as opposed to *sai api* in the original SONiC architecture).  

The *DPU/IPU/SmartNic* hardware will run a separate instance of SONiC-DASH on the hardware.  

The component interactions will be executed as a new user space container implementation; relying on the existing SONiC infrastructure and components to interact as they normally would.  

The functionality of the new *dash container* in the user space is to receive content from the Software Defined Networking (SDN) controller to control setup for the overlay configurations. DASH receives the objects, translates them with a **gNMI agent**, provides them to the *SONiC OrchAgent* for further translation onto the dataplane via the **SAI database**. 

> [!NOTE] 
> @lihuay @lguohan @prsunny - would you review and/or improve this write-up?

## DASH deployment options

The following figure is a simplified representation of DASH deployment in a datacenter. 

![dash-simplified-physical-deployment-example](images/hld/dash-simplified-physical-deployment-example-v2.svg) 

## Repositories

TBD

## References

- [Glossary](https://github.com/Azure/DASH/wiki/Glossary)

> - [SAI](../SAI)

Reviewer: https://github.com/opencomputeproject/SAI

Author: Done

- [Test](../test)

> - [SDN Features Packet Transforms](SDN-Features-Packet-Transforms.md)

Reviewer: [SAI Thrift PR](https://github.com/opencomputeproject/SAI/pull/1325)
[P4](https://opennetworking.org/p4) and [P4 working group](https://p4.org/working-groups)
[PINS](opennetworking.org/pins)
[PNA Consortium Spec](https://p4.org/p4-spec/docs/PNA-v0.5.0.html)
[IPDK](https://ipdk.io/) and [IPDK GitHub](https://github.com/ipdk-io/ipdk-io.github.io)
[bmv2 - behavioral model v2](https://github.com/p4lang/behavioral-model)
[DPDK](https://www.dpdk.org)

Author: Done

The SDN controller controls the overlay built on top of the physical layer of the infrastructure.  From the point of view of the SDN control plane, when a customer creates an operation, for example a VNET creation, from the cloud portal, the controller allocates the resources, placement management, capacity, etc. via the  **NorthBound interface APIs**.

#### SDN and DPU High-Availability (HA)

For **High Availability** (HA), the SDN controller selects the pair of cards and configures them identically.  The only requirement on the card from the HA perspective is for the cards to setup a channel between themselves for flow synchronization.  The synchronization mechanism is left for vendors to define and implement. For more information, see [High Availability and Scale]() document.   

### DASH container
 
The SDN controller communicates with a DASH device through a **[gNMI](https://github.com/Azure/DASH/wiki/Glossary#gnmi) endpoint** served by a new DASH SDN agent **running inside a new SONiC DASH container**.  

In summary:

> - The DASH container translates SDN configuration modeled in gNMI into **SONiC APPL_DB** config objects.

Reviewer: ```suggestion
- The DASH container translates SDN configuration modeled in gNMI into **SONiC DB** objects.
```

- An **enhanced syncd** will then configure the dataplane using the **vendor-specific SAI library**.

A **gNMI schema** will manage the following DASH services: 
- Elastic Network Interface (ENI)
- Access Control Lists (ACLs) 
- Routing and mappings
- Encapsulations 
- Other  

See [TODO] for DASH gNMI schema.

#### Multiple DPUs device

In the case of a multiple DPUs device the following applies:

- Each DPU provides a gNMI endpoint for SDN controller through a unique IP address. 
- An appliance or smart switch containing multiple DPUs therefore contains multiple gNMI endpoints for SDN controller, and the controller treats each DPU as a separate entity. 
- To conserve IPv4 addresses, such an appliance or switch _might_ contain a proxy (NAT) function to map a single IP address:port combination into multiple DPU endpoints with unique IPv4 addresses.  
- No complex logic will run on the switches (switches do not have a top-level view of other/neighboring switches in the infrastructure). 

### Switch State Service (SWSS)
The SWSS container comprises many daemons which operate on conceptual SONIC config objects across several databases. DASH affects the following daemons, as follows:

> * `orchagent`, which translates `APPL_DB` objects into `ASIC_DB` objects, must be enhanced to manage new DASH overlay objects, such as ACL1,2,3 rules, ENI mappings, etc. The `orchagent` has to manage both representations of SONiC objects (APPL_DB and ASIC_DB) and translates between them bidirectionally as required.

Reviewer 2: ```suggestion
* `orchagent`, which translates config/application `DB` objects into `ASIC_DB` objects, must be enhanced to manage new DASH overlay objects, such as ACL1,2,3 rules, ENI mappings, etc. The `orchagent` shall manage both representations of SONiC objects (APPL_DB and ASIC_DB).
```

Author: Fixed

## Overview 

DASH extends SONiC APIs and a comprehensive set of object models that initially describe Microsoft Azure’s networking services for the cloud. The project enlists cloud and enterprise providers to further extend DASH to meet their specific needs.

See the project [Scenario Milestone and Scoping](SDN-Features-Packet-Transforms.md#scenario-milestone-and-scoping). 

### Requirements

The overall requirement is to **optimize network SMART Programmable Technologies performance**, and **leverage commodity hardware technology** to achieve **10x or even 100x stateful connection performance**.

> - With the help of network hardware vendors, create an open forum that capitalizes on the use of **programmable networking hardware** including SmartNICs, SmartToRs, SmartAppliances. 

Reviewer: ```suggestion
- With the help of network hardware technology suppliers, create an open forum that capitalizes on the use of **programmable networking hardware** including SmartNICs, SmartToRs, SmartAppliances. 
```

> - Optimize **stateful L4** performance and connection scale by 10x or even 100x when compared to implementations that make extensive use of software. As host networking in the cloud is performed at L4 the resulting performance improvements should be truly significant.

Reviewer: ```suggestion
- Optimize **stateful L4** performance and connection scale by 10x or even 100x when compared to implementations that make extensive use of a generic software stack approach that compromises performance for flexibility.  This flexibility was needed early on, whereas the cloud is maturing and is ready for a further optimized approach.  As host networking in the cloud is performed at L4, the resulting performance improvements should be truly significant.
```

> - Microsoft Azure will integrate and deploy DASH solutions to ensure that scale, monitoring, reliability, availability and constant innovation are proven and hardened. It should be noted that innovations for **in-service software upgrades** (ISSU) and **high availability** (HA) are key tenets of the DASH charter.   

Reviewer: ```suggestion
- Microsoft Azure will integrate and deploy DASH solutions to ensure that scale, monitoring, reliability, availability and constant innovation are proven and hardened. Other enterprise and cloud providers may deploy DASH as well, and we hope to hear similar feedback and contributions as we move forward.  It should be noted that innovations for **in-service software upgrades** (ISSU) and **high availability** (HA) are key tenets of the DASH charter.   
```

**DASH architecture**

![dash-layered-architecture](images/hld/dash-layered-architecture.svg)

### SDN controller

The SDN controller is **primarily responsible for controlling the DASH overlay services**, while the traditional SONiC stack is used to manage the underlay (L3 routing) and hardware platform. 

The SDN controller controls the overlay built on top of the physical layer of the infrastructure.  From the point of view of the SDN control plane, when a customer creates an operation, for example a VNET creation, from the cloud portal, the controller allocates the resources, placement management, capacity, etc. via the  **NorthBound interface APIs**.

#### SDN and DPU High-Availability (HA)

> For **High Availability** (HA), the SDN controller selects the pair of cards and configures them identically.  The only requirement on the card from the HA perspective is for the cards to setup a channel between themselves for flow synchronization.  The synchronization mechanism is left for vendors to define and implement. For more information, see [High Availability and Scale]() document.   

Reviewer: ```suggestion
For **High Availability** (HA), the SDN controller selects the pair of cards and configures them identically.  The only requirement on the card from the HA perspective is for the cards to setup a channel between themselves for flow synchronization.  The synchronization mechanism is left for technology suppliers to define and implement. For more information, see [High Availability and Scale]() document.   
```

### DASH container
 
The SDN controller communicates with a DASH device through a **[gNMI](https://github.com/Azure/DASH/wiki/Glossary#gnmi) endpoint** served by a new DASH SDN agent **running inside a new SONiC DASH container**.  

In summary:
- The DASH container translates SDN configuration modeled in gNMI into **SONiC DB** objects.
- The **SONiC orchagent** inside the Switch State Service (SWSS) Container will be enhanced to translate these objects into **SAI_DB objects**, including the new **DASH-specific SAI objects**. 

> - An **enhanced syncd** will then configure the dataplane using the **vendor-specific SAI library**.

Reviewer: ```suggestion
- An **enhanced syncd** will then configure the dataplane using the **technology supplier-specific SAI library**.
```

#### Multiple DPUs device

In the case of a multiple DPUs device the following applies:

- Each DPU provides a gNMI endpoint for SDN controller through a unique IP address. 
- An appliance or smart switch containing multiple DPUs therefore contains multiple gNMI endpoints for SDN controller, and the controller treats each DPU as a separate entity. 
- To conserve IPv4 addresses, such an appliance or switch _might_ contain a proxy (NAT) function to map a single IP address:port combination into multiple DPU endpoints with unique IPv4 addresses.  
- No complex logic will run on the switches (switches do not have a top-level view of other/neighboring switches in the infrastructure). 

### Switch State Service (SWSS)
The SWSS container comprises many daemons which operate on conceptual SONIC config objects across several databases. DASH affects the following daemons, as follows:
* `orchagent`, which translates `xx_DB` objects into `ASIC_DB` objects, must be enhanced to manage new DASH overlay objects, such as ACL1,2,3 rules, ENI mappings, etc. The `orchagent` has to manage both representations of SONiC objects (xx_DB and ASIC_DB) and translates between them bidirectionally as required.

> * `syncd`, which translates `ASIC_DB` conceptual objects into vendor SAI library API calls, must likewise be enhanced to handle new DASH SAI objects.

Reviewer: ```suggestion
* `syncd`, which translates `ASIC_DB` conceptual objects into technology supplier SAI library API calls, must likewise be enhanced to handle new DASH SAI objects.
```

### Switch Abstraction Interface (SAI) DASH

> The Switch Abstraction Interface (SAI) is a common API that is supported by many switch ASIC vendors. SONiC uses SAI to program the ASIC. This enables SONiC to work across multiple ASIC platforms naturally. DASH uses a combination of traditional SAI headers and new DASH pipeline-specific headers. Vendors must implement this interface for their DASH devices. This is the primary integration point of DASH devices and the SONiC stack. It will be rigorously tested for performance and conformance. See [DASH Testing documentation](https://github.com/Azure/DASH/tree/main/test).

Reviewer: ```suggestion
The Switch Abstraction Interface (SAI) is a common API that is supported by many switch ASIC technology suppliers. SONiC uses SAI to program the ASIC. This enables SONiC to work across multiple ASIC platforms naturally. DASH uses a combination of traditional SAI headers and new DASH pipeline-specific headers. Technology suppliers must implement this interface for their DASH devices. This is the primary integration point of DASH devices and the SONiC stack. It will be rigorously tested for performance and conformance. See [DASH Testing documentation](https://github.com/Azure/DASH/tree/main/test).
```

SAI "schema" are represented as fixed c-language header files and derived metadata header files. The underlay and overlay schema have different origins:
* Traditional SAI headers are defined in the [OCP SAI project repo](https://github.com/opencomputeproject/SAI/tree/master/inc).These are hand-generated and maintained. DASH uses a subset of these to manage underlay functions, e.g. device management, Layer 3 routing and so forth.
* DASH SAI "underlay" objects are derived from a [P4 Behavioral Model](https://github.com/Azure/DASH/tree/main/sirius-pipeline). A script reads the P4 model and generates SAI header files.

> DASH uses an **enhanced syncd** to configure the dataplane using the vendor-specific SAI library.

Reviewer: ```suggestion
DASH uses an **enhanced syncd** to configure the dataplane using the technology supplier-specific SAI library.
```

### ASIC Drivers

> The term "ASIC Drivers" is borrowed from traditional SONiC and SAI, where a datacenter switch was implemented almost entirely inside an ASIC (or multiple ASICs). These devices are programmed using a vendor Software Development Kit (SDK) which includes device drivers, kernel modules, etc.

Reviewer: ```suggestion
The term "ASIC Drivers" is borrowed from traditional SONiC and SAI, where a datacenter switch was implemented almost entirely inside an ASIC (or multiple ASICs). These devices are programmed using a technology supplier Software Development Kit (SDK) which includes device drivers, kernel modules, etc.
```

> A contemporary DASH "SmartNIC" may consist of many complex hardware components including multi-core System On A Chip (SoC) ASICs, and the associated software. For simplicity, the software for such systems which interfaces to the SAI layer is collectively called the "ASIC driver." More importantly, the vendor SAI library will hide all details and present a uniform interface.

Reviewer: ```suggestion
A contemporary DASH "SmartNIC" may consist of many complex hardware components including multi-core System On A Chip (SoC) ASICs, and the associated software. For simplicity, the software for such systems which interfaces to the SAI layer is collectively called the "ASIC driver." More importantly, the technology supplier SAI library will hide all details and present a uniform interface.
```

  
DASH extends SONiC APIs and a comprehensive set of object models that initially describe Microsoft Azure’s networking services for the cloud.The project enlists cloud and enterprise providers to further extend DASH to meet their specific needs.

See the project [Scenario Milestone and Scoping](SDN-Features-Packet-Transforms.md#scenario-milestone-and-scoping). 

## Requirements

The overall requirement is to **optimize network SMART Programmable Technologies performance**, and **leverage commodity hardware technology** to achieve **10x or even 100x stateful connection performance**.
- With the help of network hardware vendors, create an open forum that capitalizes on the use of **programmable networking hardware** including SmartNICs, SmartToRs, SmartAppliances. 
- Optimize **stateful L4** performance and connection scale by 10x or even 100x when compared to implementations that make extensive use of software. As host networking in the cloud is performed at L4 the resulting performance improvements should be truly significant.

> - Microsoft Azure will integrate and deploy DASH solutions to ensure that scale, monitoring, reliability, availability and constant innovation are proven and hardened. It should be noted that innovations for **in-service software upgrades** (ISSU) and **high availability** (HA) are key tenets of the DASH charter that will also be battle hardened. 

Reviewer: ```suggestion
- Microsoft Azure will integrate and deploy DASH solutions to ensure that scale, monitoring, reliability, availability and constant innovation are proven and hardened. It should be noted that innovations for **in-service software upgrades** (ISSU) and **high availability** (HA) are key tenets of the DASH charter.   
```

## Architecture

SONiC is structured into various containers that communicate through multiple logical databases via a shared Redis instance. DASH will make use of the SONiC infrastructure as shown in the figure below.

> [!NOTE] 
> **The figures shown are a work in progress**

### DASH NOS single DPU on NIC

![dash-single-dpu-architecture](images/hld/dash-single-dpu-architecture.svg)

### DASH appliance architecture

#### High level architecture

![dash-high-level-appliance](images/hld/dash-high-level-appliance.svg)

#### Low level architecture

![dash-appliance-architecture](images/hld/dash-appliance-architecture.svg)

### DASH smart switch architecture

#### High level architecture

![dash-high-level-smart-switch](images/hld/dash-high-level-smart-switch.svg)

#### Low level architecture

![dash-smart-switch-architecture](images/hld/dash-smart-switch-architecture.svg)

## High-Level Design

**DASH high level design**

> 

Reviewer: The system architecture for SONiC-DASH relies upon the [SONiC system architecture](https://github.com/Azure/SONiC/wiki/Architecture) and adds a _new docker container_ in the user space we are naming 'dash container' to create the functional component for DASH.  
In the sync-d container we will add 'sai api DASH' (as opposed to simply 'sai api' in the original SONiC architecture.  
The DPU/IPU/SmartNic hardware will run a separate instance of 'SONiC-DASH' on the hardware.  
The component interactions will be executed as a new user space container implementation; relying on the existing SONiC infrastructure and components to interact as they normally would.  
The functionality of the new dash container in the user space is to receive content from the Software Defined Networking (SDN) controller to control setup for the overlay configurations.  DASH receives the objects, translates them with a gNMI agent, provides them to the SONiC OrchAgent for further translation onto the dataplane via the SAI db. 

@lguohan @prsunny  - we need a drawing and descriptions of the DASH user space **state interactions** figure (example below for [LLDP-state interactions](https://github.com/Azure/SONiC/wiki/Architecture)

![image](https://user-images.githubusercontent.com/22226040/150199558-9386569e-cffe-4462-9af3-43f4f20b1dab.png)

Reviewer 3: This is very useful and concise.

Author: Accepted.

- With the help of network hardware vendors, create an open forum that capitalizes on the use of **programmable networking hardware** including SmartNICs, SmartToRs, SmartAppliances. 
- Optimize **stateful L4** performance and connection scale by 10x or even 100x when compared to implementations that make extensive use of software. As host networking in the cloud is performed at L4 the resulting performance improvements should be truly significant.
- Microsoft Azure will integrate and deploy DASH solutions to ensure that scale, monitoring, reliability, availability and constant innovation are proven and hardened. It should be noted that innovations for **in-service software upgrades** (ISSU) and **high availability** (HA) are key tenets of the DASH charter that will also be battle hardened. 

## Architecture

SONiC is structured into various containers that communicate through multiple logical databases via a shared Redis instance. DASH will make use of the SONiC infrastructure as shown in the figure below.

> [!NOTE] 
> **The figures shonw a work in progress**

### DASH NOS single DPU on NIC

![dash-single-dpu-architecture](images/hld/dash-single-dpu-architecture.svg)

> ### DASH smart switch architecture

Reviewer: Could we switch the order to be DPU, High Level Appliance, Low Level Appliance, High Level Smart Switch, Low Level Smart Switch ?

Author: Done

#### High level architecture

![dash-high-level-smart-switch](images/hld/dash-high-level-smart-switch.svg)

#### Low level architecture

![dash-smart-switch-architecture](images/hld/dash-smart-switch-architecture.svg)

### DASH appliance architecture

> 

Reviewer: Could we add the 'High Level Architecture' heading here?

Author: Done

> 

Reviewer: ```suggestion
I think we need to add #### High Level Architecture here, it seems to be missing
```

Author: Added

### Switch State Service (SWSS)
The SWSS container comprises many daemons which operate on conceptual SONIC config objects across several databases. DASH affects the following daemons, as follows:

> * `orchagent`, which translates `DB` objects into `ASIC_DB` objects, must be enhanced to manage new DASH overlay objects, such as ACL1,2,3 rules, ENI mappings, etc. The `orchagent` has to manage both representations of SONiC objects (APPL_DB and ASIC_DB) and translates between them bidirectionally as required.

Reviewer 3: The two sentences should be consistent in the use of `DB` and `APPL_DB`. How about:

`orchagent`, which translates various application-layer `xx_DB` objects (e.g, `APPL_DB`) into abstract dataplane-layer `ASIC_DB` objects, must be enhanced to manage new DASH overlay objects, such as ACL1,2,3 rules, ENI mappings, etc. The `orchagent` has to manage both representations of SONiC objects (`xx_DB` and `ASIC_DB`) and translates between them bidirectionally, as required.

Author: Fixed.
